volsupervisor/loop.go

```diff
@@ -28,48 +56,48 @@ func (dc *DaemonConfig) pruneSnapshots(val *config.Volume) {
 		Reason: lock.ReasonSnapshotPrune,
 	}
 
-	err := lock.NewDriver(dc.Config).ExecuteWithUseLock(uc, func(ld *lock.Driver, uc config.UseLocker) error {
-		driver, err := backend.NewSnapshotDriver(val.Backends.Snapshot)
-		if err != nil {
-			logrus.Errorf("failed to get driver: %v", err)
-			return err
-		}
+	stopChan, err := lock.NewDriver(dc.Config).AcquireWithTTLRefresh(uc, dc.Global.TTL, dc.Global.Timeout)
+	if err != nil {
+		logrus.Error(errors.LockFailed.Combine(err))
+		return
+	}
 
-		driverOpts := storage.DriverOptions{
-			Volume: storage.Volume{
-				Name: val.String(),
-				Params: storage.Params{
-					"pool": val.DriverOptions["pool"],
-				},
-			},
-			Timeout: dc.Global.Timeout,
-		}
+	defer func() { stopChan <- struct{}{} }()
 
-		list, err := driver.ListSnapshots(driverOpts)
-		if err != nil {
-			logrus.Errorf("Could not list snapshots for volume %q: %v", val.VolumeName, err)
-			return err
-		}
+	driver, err := backend.NewSnapshotDriver(val.Backends.Snapshot)
+	if err != nil {
+		logrus.Errorf("failed to get driver: %v", err)
+		return
+	}
 
-		logrus.Debugf("Volume %q: keeping %d snapshots", val, val.RuntimeOptions.Snapshot.Keep)
+	driverOpts := storage.DriverOptions{
+		Volume: storage.Volume{
+			Name: val.String(),
+			Params: storage.Params{
+				"pool": val.DriverOptions["pool"],
+			},
+		},
+		Timeout: dc.Global.Timeout,
+	}
 
-		toDeleteCount := len(list) - int(val.RuntimeOptions.Snapshot.Keep)
-		if toDeleteCount < 0 {
-			return nil
-		}
+	list, err := driver.ListSnapshots(driverOpts)
+	if err != nil {
+		logrus.Errorf("Could not list snapshots for volume %q: %v", val.VolumeName, err)
+		return
+	}
 
-		for i := 0; i < toDeleteCount; i++ {
-			logrus.Infof("Removing snapshot %q for volume %q", list[i], val.VolumeName)
-			if err := driver.RemoveSnapshot(list[i], driverOpts); err != nil {
-				logrus.Errorf("Removing snapshot %q for volume %q failed: %v", list[i], val.VolumeName, err)
-			}
-		}
+	logrus.Debugf("Volume %q: keeping %d snapshots", val, val.RuntimeOptions.Snapshot.Keep)
 
-		return nil
-	})
+	toDeleteCount := len(list) - int(val.RuntimeOptions.Snapshot.Keep)
+	if toDeleteCount < 0 {
+		return
+	}
 
-	if err != nil {
-		logrus.Errorf("Error removing snapshot for volume %q: %v", val, err)
+	for i := 0; i < toDeleteCount; i++ {
+		logrus.Infof("Removing snapshot %q for volume %q", list[i], val.VolumeName)
+		if err := driver.RemoveSnapshot(list[i], driverOpts); err != nil {
+			logrus.Errorf("Removing snapshot %q for volume %q failed: %v", list[i], val.VolumeName, err)
+		}
 	}
 }
```

Comment:
read line 67 please, and the code in lock/lock.go for AcquireWithTTLRefresh. Pay specific attention to how `stopChan` is used.